weblate/utils/ratelimit.py

```diff
@@ -159,3 +109,120 @@ def _rate_wrap(
         return wraps(function)(_rate_wrap)
 
     return _session_ratelimit_post_controller
+
+
+def rate_limit_notify(address: str) -> tuple[bool, str]:
+    """
+    Multi-level rate limiting for email notifications.
+
+    Returns: tuple: (is_blocked, reason)
+    """
+    rate_limits = getattr(settings, "RATELIMIT_NOTIFICATION_LIMITS", [(1000, 86400)])
+    encoded_email = siphash("Weblate notifier", address)
+    limiter = RateLimitNotify(f"{encoded_email}", rate_limits)
+    return limiter.is_limit_exceeded()
+
+
+class RateLimitBase:
+    cache_items: list[CacheCounterItem]
+    key: str
+
+    def __init__(self, base_key: str, rate_limits: list[tuple[int, int]]):
+        self.key = base_key
+
+        self.cache_items = [
+            CacheCounterItem(self.key, attempts, window)
+            for attempts, window in rate_limits
+        ]
+
+    def is_limit_exceeded(self) -> tuple[bool, str]:
+        # Check all without decrementing
+        for cache_item in self.cache_items:
+            if cache_item.count_remaining <= 0:
```

Comment:
This is where the test with RATELIMIT_MESSAGE_ATTEMPTS=1, RATELIMIT_WINDOW=0 fails.